src/main/CommandLine.cpp

```diff
@@ -964,6 +965,78 @@ runSimulate(CommandLineArgs const& args)
         });
 }
 
+int
+runSimulateBuckets(CommandLineArgs const& args)
+{
+    CommandLine::ConfigOption configOption;
+    uint32_t ledger = 0;
+    uint32_t n = 0;
+
+    ParserWithValidation firstLedgerParser{
+        clara::Opt{ledger, "LEDGER"}["--ledger"]("ledger to get bucket state")};
+
+    return runWithHelp(
+        args,
+        {configurationParser(configOption), firstLedgerParser,
+         clara::Opt{n, "N"}["--multiplier"]("amplification factor")},
+        [&] {
+            auto config = configOption.getConfig();
+            config.setNoListen();
+            // Redirect all the bucket adoption to special directory to not mess
+            // up our current buckets
+            config.BUCKET_DIR_PATH = "simulate-buckets";
+
+            VirtualClock clock(VirtualClock::REAL_TIME);
+            auto app = Application::create(clock, config, true);
+            app->start();
+            TmpDir dir(app->getTmpDirManager().tmpDir("simulate-buckets-tmp"));
+
+            auto checkpoint =
+                ledger == CatchupConfiguration::CURRENT
+                    ? CatchupConfiguration::CURRENT
+                    : app->getHistoryManager().checkpointContainingLedger(
+                          ledger);
+
+            auto simulateBuckets = std::make_shared<SimulateApplyBucketsWork>(
+                *app, n, checkpoint, dir);
+
+            // Once simulated bucketlist is good to go, download ledgers headers
+            // to convince LedgerManager that we have succesfully restored
+            // ledger state
+            LedgerRange lr{checkpoint, checkpoint};
+            CheckpointRange cr{lr, app->getHistoryManager()};
+            auto downloadHeaders = std::make_shared<BatchDownloadWork>(
+                *app, cr, HISTORY_FILE_TYPE_LEDGER, dir);
+
+            std::vector<std::shared_ptr<BasicWork>> seq{simulateBuckets,
+                                                        downloadHeaders};
+
+            auto w = app->getWorkScheduler().executeWork<WorkSequence>(
+                "simulate-bucketlist-application", seq);
+
+            if (w->getState() == BasicWork::State::WORK_SUCCESS)
+            {
+                // Ensure that LedgerManager's view of LCL if consistent with
+                // the bucketlist
+                FileTransferInfo ft(dir, HISTORY_FILE_TYPE_LEDGER, checkpoint);
+                XDRInputFileStream hdrIn;
+                hdrIn.open(ft.localPath_nogz());
+                LedgerHeaderHistoryEntry curr;
+                while (hdrIn && hdrIn.readOne(curr))
+                    ;
```

Comment:
Is this loop supposed to be empty? If so, then I don't really understand what this is doing so perhaps a comment would be helpful.

Reply:
yes. it's reading the last ledger header